File: problems/java/MeetingScheduler/src/services/MeetingScheduler.java
package services;

import enums.NotificationType;
import enums.RSVPStatus;
import models.*;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;

public class MeetingScheduler {

    private static volatile MeetingScheduler instance;
    private static final Object lock = new Object();

    private final NotificationService notificationService;
    private final List<MeetingRoom> meetingRooms;

    private MeetingScheduler() {
        meetingRooms = new ArrayList<>();
        notificationService = new NotificationService();
    }

    public static MeetingScheduler getInstance() {
        if (instance == null) {
            synchronized (lock) {
                if (instance == null) {
                    instance = new MeetingScheduler();
                }
            }
        }
        return instance;
    }

    public void addRooms(List<MeetingRoom> rooms) {
        meetingRooms.addAll(rooms);
        meetingRooms.sort(Comparator.comparingInt(MeetingRoom::getCapacity));
    }

    public Optional<Meeting> scheduleMeeting(User organizer, List<User> participants, Interval interval, String subject) {

        int capacity = participants.size() + 1; // Participants + organizer
        MeetingRoom room = bookMeetingRoom(interval, capacity);
        if (room == null) {
            System.out.println("Meeting Room not available at the given interval to schedule the meeting");
            return Optional.empty();
        }

        Meeting meeting = new Meeting(subject, organizer, participants, room, interval);

        for (User participant : participants) {
            participant.addMeetingToCalendar(meeting);
        }

        organizer.addMeetingToCalendar(meeting);
        organizer.respondInvitation(meeting, RSVPStatus.ACCEPTED);

        notificationService.sendNotifications(NotificationType.INVITE, participants, meeting);

        System.out.println("Meeting Scheduled Successfully");

        return Optional.of(meeting);
    }

    public void cancelMeeting(Meeting meeting) {
        if (!meeting.isActive()) {
            System.out.println("Meeting already cancelled");
            return;
        }
        meeting.setActive(false);

        MeetingRoom room = meeting.getBookedRoom();
        Interval interval = meeting.getBookedInterval();
        room.releaseInterval(interval);

        List<User> participants = meeting.getAllParticipants();
        for (User attendee : participants) {
            attendee.removeMeetingFromCalendar(meeting);
        }
        meeting.getOrganizer().removeMeetingFromCalendar(meeting);

        notificationService.sendNotifications(NotificationType.CANCEL, participants, meeting);

        System.out.println("Meeting Cancelled Successfully");
    }

    public void addParticipants(Meeting meeting, List<User> participants) {
        if (!meeting.isActive()) {
            System.out.println("Cannot add participants — meeting is cancelled");
            return;
        }
        for (User participant : participants) {
            meeting.addParticipant(participant);
            participant.addMeetingToCalendar(meeting);
        }

        notificationService.sendNotifications(NotificationType.INVITE, participants, meeting);

        System.out.println("Participants Added Successfully");
    }

    public void removeParticipants(Meeting meeting, List<User> participants) {
        if (!meeting.isActive()) {
            System.out.println("Cannot remove participants — meeting is cancelled");
            return;
        }
        for (User participant : participants) {
            meeting.removeParticipant(participant);
            participant.removeMeetingFromCalendar(meeting);
        }

        notificationService.sendNotifications(NotificationType.CANCEL, participants, meeting);

        System.out.println("Participants Removed Successfully");
    }

    public List<User> getParticipantsWithConflicts(List<User> participants, Interval interval) {

        List<User> conflictedUsers = new ArrayList<>();
        for (User candidate : participants) {
            if (candidate.hasConflict(interval)) {
                conflictedUsers.add(candidate);
            }
        }
        return conflictedUsers;
    }

    private MeetingRoom bookMeetingRoom(Interval interval, int capacity) {
        for (MeetingRoom room : meetingRooms) {
            if (room.checkAndBookInterval(capacity, interval)) {
                System.out.println("Room : " + room.getId() + " booked successfully");
                return room;
            }
        }
        return null;
    }
}

File: problems/python/MeetingScheduler/src/services/meeting_scheduler.py
from threading import Lock
from typing import Optional

from enums.notification_type import NotificationType
from enums.rsvp_status import RSVPStatus
from models.interval import Interval
from models.meeting import Meeting
from models.meeting_room import MeetingRoom
from models.user import User
from services.notification_service import NotificationService


class MeetingScheduler:
    _instance: Optional["MeetingScheduler"] = None
    _instance_lock = Lock()

    def __init__(self) -> None:
        self._meeting_rooms: list[MeetingRoom] = []
        self._rooms_lock = Lock()
        self._notification_service = NotificationService()

    @classmethod
    def get_instance(cls) -> "MeetingScheduler":
        if cls._instance is None:
            with cls._instance_lock:
                if cls._instance is None:
                    cls._instance = cls()
        return cls._instance

    def add_rooms(self, rooms: list[MeetingRoom]) -> None:
        with self._rooms_lock:
            self._meeting_rooms.extend(rooms)
            self._meeting_rooms.sort(key=lambda room: room.capacity)

    def schedule_meeting(
        self, organizer: User, participants: list[User], interval: Interval, subject: str
    ) -> Optional[Meeting]:
        capacity = len(participants) + 1
        room = self._book_meeting_room(interval, capacity)
        if room is None:
            print("Meeting Room not available at the given interval to schedule the meeting")
            return None

        meeting = Meeting(subject, organizer, participants, room, interval)
        for participant in participants:
            participant.add_meeting_to_calendar(meeting)

        organizer.add_meeting_to_calendar(meeting)
        organizer.respond_invitation(meeting, RSVPStatus.ACCEPTED)

        self._notification_service.send_notifications(
            NotificationType.INVITE, participants, meeting
        )
        print("Meeting Scheduled Successfully")
        return meeting

    def cancel_meeting(self, meeting: Meeting) -> None:
        if not meeting.active:
            print("Meeting already cancelled")
            return

        meeting.active = False
        meeting.booked_room.release_interval(meeting.booked_interval)
        participants = meeting.get_all_participants()
        for attendee in participants:
            attendee.remove_meeting_from_calendar(meeting)
        meeting.organizer.remove_meeting_from_calendar(meeting)

        self._notification_service.send_notifications(
            NotificationType.CANCEL, participants, meeting
        )
        print("Meeting Cancelled Successfully")

    def add_participants(self, meeting: Meeting, participants: list[User]) -> None:
        if not meeting.active:
            print("Cannot add participants — meeting is cancelled")
            return

        for participant in participants:
            meeting.add_participant(participant)
            participant.add_meeting_to_calendar(meeting)

        self._notification_service.send_notifications(
            NotificationType.INVITE, participants, meeting
        )
        print("Participants Added Successfully")

    def remove_participants(self, meeting: Meeting, participants: list[User]) -> None:
        if not meeting.active:
            print("Cannot remove participants — meeting is cancelled")
            return

        for participant in participants:
            meeting.remove_participant(participant)
            participant.remove_meeting_from_calendar(meeting)

        self._notification_service.send_notifications(
            NotificationType.CANCEL, participants, meeting
        )
        print("Participants Removed Successfully")

    @staticmethod
    def get_participants_with_conflicts(
        participants: list[User], interval: Interval
    ) -> list[User]:
        conflicted_users: list[User] = []
        for candidate in participants:
            if candidate.has_conflict(interval):
                conflicted_users.append(candidate)
        return conflicted_users

    def _book_meeting_room(self, interval: Interval, capacity: int) -> Optional[MeetingRoom]:
        with self._rooms_lock:
            rooms = list(self._meeting_rooms)
        for room in rooms:
            if room.check_and_book_interval(capacity, interval):
                print(f"Room : {room.id} booked successfully")
                return room
        return None
